src/service/viewer-impl.js

```diff
@@ -499,6 +520,56 @@ export class Viewer {
     return this.isWebviewEmbedded_;
   }
 
+  /**
+   * Whether the document is embedded in a Chrome Custom Tab.
+   * @return {boolean}
+   */
+  isCctEmbedded() {
+    return this.isCctEmbedded_;
+  }
+
+  /**
+   * Update the URL fragment with data needed to support custom tabs. This will
+   * not clear query string parameters, but will clear the fragment.
+   */
+  maybeUpdateFragmentForCct() {
+    if (!this.isCctEmbedded_) {
+      return;
+    }
+    // CCT only works with versions of Chrome that support the history API.
+    if (!this.win.history.replaceState) {
+      return;
+    }
+    const sourceOrigin = getSourceOrigin(this.win.location.href);
+    const canonicalUrl = Services.documentInfoForDoc(this.ampdoc).canonicalUrl;
+    const canonicalSourceOrigin = getSourceOrigin(canonicalUrl);
+    if (this.hasRoughlySameOrigin_(sourceOrigin, canonicalSourceOrigin)) {
+      const oldFragment = getFragment(this.win.location.href);
```

Comment:
Use `originalHash`, the fragment in `location.href` will be blown away at startup.